lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,294 @@
+#include "seahorn/HornClauseDB.hh"
+
+#include "llvm/IR/Function.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "ufo/Expr.hpp"
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include <vector>
+
+using namespace llvm;
+
+namespace seahorn
+{
+  char Houdini::ID = 0;
+
+  struct IsBVar : public std::unary_function<Expr, bool>
+  {
+     IsBVar () {}
+     bool operator() (Expr e)
+     {return bind::isBVar (e);}
+  };
+
+  struct IsPredApp : public std::unary_function<Expr, bool>
+  {
+     HornClauseDB &m_db;
+     IsPredApp (HornClauseDB &db) : m_db (db) {}
+
+     bool operator() (Expr e)
+     {return bind::isFapp (e) && m_db.hasRelation (bind::fname(e));}
+  };
+
+  /*
+   * Extract all bvars in an expression
+   */
+  template<typename OutputIterator>
+  void Houdini::get_all_bvars (Expr e, OutputIterator out)
+  {filter (e, IsBVar(), out);}
+
+  /*
+   * Extract all predicates in an expression
+   */
+  template<typename OutputIterator>
+  void Houdini::get_all_pred_apps (Expr e, HornClauseDB &db, OutputIterator out)
+  {filter (e, IsPredApp(db), out);}
+
+  bool Houdini::runOnModule (Module &M)
+  {
+    HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+    //load the Horn clause database
+    auto &db = hm.getHornClauseDB ();
+    db.buildIndexes ();
+
+    runHoudini(db, hm);
+
+    return false;
+  }
+
+  void Houdini::getAnalysisUsage (AnalysisUsage &AU) const
+  {
+    AU.addRequired<HornifyModule> ();
+    AU.setPreservesAll();
+  }
+
+  /*
+   * Use template to convert a predicate to candidate
+   */
+  Expr Houdini::relToCand(Expr fdecl)
+  {
+	ExprVector bvars;
+	ExprVector bins;   // a vector of LT expressions
+	Expr cand = NULL;
+
+	int bvar_count = 0;
+	unsigned i = 0;
+	for (i=0; i < bind::domainSz(fdecl); i++)
+	{
+		// if its type is INT
+		if (isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+		{
+			// what is efac?
+			Expr bvar = bind::bvar (i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac())); //the id of bvar is the same as related arg index
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+
+	return cand;
+  }
+
+  #define SAT true
+  #define UNSAT false
+
+  std::map<Expr, Expr> Houdini::currentCandidates;	//a map from predicates to candidates
```

Comment:
debug it. why do you think this happens when map is destroyed. is
expression factory destroyed before all expressions are?

look at similar code. what does HornSolver do with expressions it creates.
Are there other passes that keep expressions around, etc.

cheers
arie
On Tue, Jun 28, 2016 at 18:28 Chenguang-Zhu <notifications@github.com>
wrote:

> In lib/seahorn/Houdini.cc
> <https://github.com/seahorn/seahorn/pull/33#discussion_r68855939>:
>
> > +	{
> > +		for(int j=0; j<bvars.size()-1; j++)
> > +		{
> > +			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
> > +			bins.push_back(lt);
> > +		}
> > +		cand = mknary<AND>(bins.begin(), bins.end());
> > +	}
> > +
> > +	return cand;
> > +  }
> > +
> > +  #define SAT true
> > +  #define UNSAT false
> > +
> > +  std::map<Expr, Expr> Houdini::currentCandidates;	//a map from predicates to candidates
>
> I tried not to use static field. Instead, using
>
> private:
> std::map currentCandidates;
>
> in Houdini.hh file.
> But it causes error when destroy the map. It shows:
>
> Program received signal SIGSEGV, Segmentation fault.
> ~intrusive_ptr (this=0x1f263e8, __in_chrg=) at
> /usr/local/boost_1_61_0/boost/smart_ptr/intrusive_ptr.hpp:97
> 97 if( px != 0 ) intrusive_ptr_release( px );
>
> What should I do with this error?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seahorn/seahorn/pull/33/files/d6d85bf00b64f3e51e908548e3ce060196ed627b#r68855939>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFzN59kABvSakdi-4EUs8CJS1dME3JjLks5qQaAdgaJpZM4I7V-j>
> .
>